src/ripple/app/misc/impl/ValidatorList.cpp

```diff
@@ -191,33 +200,164 @@ ValidatorList::load (
+        value["blob"] = publisher.rawBlob;
+        value["signature"] = publisher.rawSignature;
+        value["version"] = publisher.rawVersion;
+
+        writeFileContents(ec, filename, value.toStyledString());
+
+        if (ec)
+        {
+            // Log and ignore any file I/O exceptions
+            JLOG(j_.error()) <<
+                "Problem writing " <<
+                filename <<
+                " " <<
+                ec.value() <<
+                ": " <<
+                ec.message();
+        }
+    }
+}
+
+ValidatorList::PublisherListStats
+ValidatorList::applyListAndBroadcast(
+    std::string const& manifest,
+    std::string const& blob,
+    std::string const& signature,
+    std::uint32_t version,
+    std::string siteUri,
+    uint256 const& hash,
+    Overlay& overlay,
+    HashRouter& hashRouter)
+{
+    auto const result = applyList(manifest, blob, signature,
+        version, std::move(siteUri), hash);
+    auto const disposition = result.disposition;
+
+    bool broadcast = disposition == ListDisposition::accepted ||
+        disposition == ListDisposition::same_sequence;
+
+    if (broadcast)
+    {
+        assert(result.available && result.publisherKey && result.sequence);
+        auto const toSkip = hashRouter.shouldRelay(hash);
+
+        if (toSkip)
+        {
+            protocol::TMValidatorList msg;
+            msg.set_manifest(manifest);
+            msg.set_blob(blob);
+            msg.set_signature(signature);
+            msg.set_version(version);
+
+            auto const& publisherKey = *result.publisherKey;
+            auto const sequence = *result.sequence;
+
+            auto peerNoSendCondition =
+                [&publisherKey, &sequence]
+                (std::shared_ptr<Peer> const& peer)
+                {
+                    return !peer->supportsProtocolVersion("XRPL/2.1") ||
+                        peer->publisherListSequence(publisherKey) >= sequence;
+                };
+            auto peerupdate =
+                [&publisherKey, sequence, &hashRouter, &hash, &j =j_]
+                (std::shared_ptr<Peer>& peer)
+                {
+                    JLOG(j.debug()) <<
+                        "Sent validator list for " <<
+                        strHex(publisherKey) << " with sequence " <<
+                        sequence << " to " <<
+                        peer->getRemoteAddress().to_string() <<
+                        " (" << peer->id() << ")";
+                    // Don't send it next time.
+                    hashRouter.addSuppressionPeer(
+                        hash, peer->id());
+                    peer->setPublisherListSequence(
+                        publisherKey, sequence);
+                };
+            overlay.foreach_mutable(send_and_modify_if_not(
+                std::make_shared<Message>(msg, protocol::mtVALIDATORLIST),
+                peer_in_set_or_condition(*toSkip, peerNoSendCondition),
+                peerupdate));
```

Comment:
I noticed there was some previous conversation that led you to remove some callbacks elsewhere. Can we remove these too? `foreach_mutable`, which you added, just calls `getActivePeers`, which is a public method. Then we can get rid of the non-lambda function objects that end up splitting the body of this loop across several files with lots of distracting boilerplate cluttered around the juicy bits.

Reply:
Nice catch. I got caught up in the copypasta, and didn't even think to check the visibility of `getActivePeers`.